### PR TITLE
Support union types, increase psalm level to 2, remove `ValidatableModelInterface`, move config to web from common, improvements

### PR DESCRIPTION
**Pull request by rustamwin**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️/❌
| Fixed issues  | #27

### CODE REVIEW

File: psalm.xml [hunk 1]
```diff
@@ -1,15 +1,11 @@
...
-    errorLevel="6"
-    resolveFromConfigFile="true"
+    errorLevel="2"
     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     xmlns="https://getpsalm.org/schema/config"
     xsi:schemaLocation="https://getpsalm.org/schema/config vendor/vimeo/psalm/config.xsd"
 >
     <projectFiles>
         <directory name="src" />
-        <ignoreFiles>
```
Comment (vjik) [inline psalm.xml:11]: Ignore vendor is userful. In some cases psalm follow from package code to vendor and check him.

File: src/RequestModelFactory.php [hunk 2]
```diff
@@ -27,17 +30,19 @@ public function __construct(ValidatorInterface $validator, Injector $injector, C
...
      * @return array
      */
-    public function createInstances(ServerRequestInterface $request, array $handlerParams): array
+    public function createInstances(ServerRequestInterface $request, array $handlerParameters): array
     {
         $requestModelInstances = [];
-        foreach ($this->getModelRequestClasses($handlerParams) as $modelClass) {
-            $requestModelInstances[] = $this->processModel($request, $this->injector->make($modelClass));
+        foreach ($this->getModelRequestClasses($handlerParameters) as $modelClass) {
+            /** @var RequestModelInterface $modelInstance */
```
Comment (vjik) [inline src/RequestModelFactory.php:43]: Try in `getModelRequestClasses()` returns `class-string<RequestModelInterface>[]`. Maybe this phpdoc not need.

File: psalm.xml (hunk 1 above)
Comment (rustamwin) [reply psalm.xml]: Alright

File: src/RequestModelFactory.php (hunk 2 above)
Comment (rustamwin) [reply src/RequestModelFactory.php:43]: This phpdoc for `$modelInstance` not `$modelClass`

Comment (vjik) [reply src/RequestModelFactory.php:43]: Yes, but injector support typed return for psalm: https://github.com/yiisoft/injector/blob/d936d15e8a0c70d3486832ad75edf73688d4c309/src/Injector.php#L107

File: src/RequestModelFactory.php [hunk 3]
```diff
@@ -58,29 +62,60 @@ private function processModel(ServerRequestInterface $request, RequestModelInter
...
+     *
+     * @psalm-suppress UndefinedClass
+     *
+     * @throws ReflectionException
+     *
+     * @return bool
+     */
+    private function parameterIsRequestModel(
+        ReflectionParameter $parameter,
+        ReflectionType &$parameterType = null
```
Comment (vjik) [inline src/RequestModelFactory.php:98]: Second argument always used, we can remove `= null`.

**Review by vjik — COMMENTED** [review]
Is there a case when union type is needed?
Maybe for union type throw exception with an explanation that only single type should be used?

**Comment by rustamwin** [comment]
> Is there a case when union type is needed?

I have no any case.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Hmm "I have no any case" is the problem.

**Comment by samdark** [comment]
Closing this one since there seems to be no valid use-case for union types.

**Pull request closed without merging**